src/util/contracthash.rs

```diff
@@ -85,23 +81,89 @@ impl error::Error for Error {
             Error::ExpectedKey => "expected key when deserializing script",
             Error::ExpectedChecksig => "expected OP_*CHECKSIG* when deserializing script",
             Error::TooFewKeys(_) => "too few keys for template",
-            Error::TooManyKeys(_) => "too many keys for template"
+            Error::TooManyKeys(_) => "too many keys for template",
+            Error::WrongScriptPubkeyType(_) => "unsupported scriptPubkey type"
         }
     }
+
+    fn cause(&self) -> Option<&error::Error> {
+        match *self {
+            Error::Secp(ref e) => Some(e),
+            Error::Script(ref e) => Some(e),
+            _ => None
+        }
+    }
+}
+
+/// Version used by `scriptPubKey` with SegWit support
+#[derive(Debug, Display, Clone, Copy, Eq, PartialEq)]
+pub enum WitnessScriptVersion {
+    /// Legacy P2SH-wrapped representation for witnessScript
+    LegacyP2SH = -1,
+    /// V0 of witnessScript
+    V0 = 0
+}
+
+/// Transaction `scriptPubKey` formats
+#[derive(Debug, Display, Clone, Eq, PartialEq)]
+pub enum ScriptPubkeyType {
+    /// Pay-to-public key (legacy, deprecated)
+    P2PK,
+    /// Pay-to-public key hash
+    P2PKH,
+    /// Pay-to-witness public key hash (including P2SH wrapped version)
+    P2WPKH(WitnessScriptVersion),
```

Comment:
This is not correct, right? There is only one scrptPubKey type for p2wpkh, it's native p2wpkh. A p2shwpkh is just a p2sh. p2shwpkh is not really a script type, nor an address type; it's more like a script derivation strategy.

Reply:
While the PR is closed and it is decided to move these efforts to a wallet library, I'd like to finish this discussion on the nature of P2SH-P2WPK/SH, in order to be on the same page when we will be reviewing this functionality in the other library.

I would not agree that P2SH-P2WPK/SH is just a script derivation strategy, since transactions spending such outputs would be rejected by the non-SegWit nodes, i.e. they do differ from normal P2SH. The scriptPubKey in tx out is really indifferentiable, while in tx input spendig such output sigScript will have a different redeemScript structure plus a special witnessScript as for any SegWit transaction.

Reply:
Non-segwit nodes will accept everything that segwit nodes accept, and then some.

Reply:
Right, would it be more appropriate to say that they do accept them due to a different reason than SegWit-enabled, i.e. the scripts in tx in are of different (i.e. "distinguishably different") content?